crates/bevy_ecs/src/event/mod.rs

```diff
@@ -1,39 +1,322 @@
+/// the hierarchy (from child to parent).
+///
+/// You can also specify your own [`Traversal`](crate::traversal::Traversal) implementation. A common pattern is to use
+/// [`Relationship`](crate::relationship::Relationship) components, which will follow the relationships to their root
+/// (just be sure to avoid cycles ... these aren't detected for performance reasons):
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Component)]
+/// #[relationship(relationship_target = ClickableBy)]
+/// struct Clickable(Entity);
+
+/// #[derive(Component)]
+/// #[relationship_target(relationship = Clickable)]
+/// struct ClickableBy(Vec<Entity>);
+///
+/// #[derive(EntityEvent)]
+/// #[entity_event(propagate = &'static Clickable)]
+/// struct Click {
+///     entity: Entity,
+/// }
+/// ```
+///
+/// By default, propagation requires observers to opt-in:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(EntityEvent)]
+/// #[entity_event(propagate)]
+/// struct Click {
+///     entity: Entity,
+/// }
+///
+/// # let mut world = World::default();
+/// world.add_observer(|mut click: On<Click>| {
+///   // this will propagate the event up to the parent, using `ChildOf`
+///   click.propagate(true);
+/// });
+/// ```
+///
+/// But you can enable auto propagation using the `#[entity_event(auto_propagate)]` attribute:
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(EntityEvent)]
+/// #[entity_event(propagate, auto_propagate)]
+/// struct Click {
+///     entity: Entity,
+/// }
+/// ```
+///
+/// ## Manually spawning [`EntityEvent`] observers
+///
+/// The examples above that call [`EntityWorldMut::observe`] to add entity-specific observer logic are
+/// just shorthand for spawning an [`Observer`] directly and manually watching the entity:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # let mut world = World::default();
+/// # let entity = world.spawn_empty().id();
+/// # #[derive(EntityEvent)]
+/// # struct Explode(Entity);
+/// let mut observer = Observer::new(|event: On<Explode>| {});
+/// observer.watch_entity(entity);
+/// world.spawn(observer);
+/// ```
+///
+/// Note that the [`Observer`] component is not added to the entity it is observing. Observers should always be their own entities, as there
+/// can be multiple observers of the same entity!
+///
+/// You can call [`Observer::watch_entity`] more than once or [`Observer::watch_entities`] to watch multiple entities with the same [`Observer`].
+///
+/// [`EntityWorldMut::observe`]: crate::world::EntityWorldMut::observe
+/// [`Observer`]: crate::observer::Observer
+/// [`Observer::watch_entity`]: crate::observer::Observer::watch_entity
+/// [`Observer::watch_entities`]: crate::observer::Observer::watch_entities
+pub trait EntityEvent: Event {
+    /// The [`Entity`] "target" of this [`EntityEvent`]. When triggered, this will run observers that watch for this specific entity.
```

Comment:
Calling out some conventions around "this should be a struct field" and "use meaningful names" and "it's probably clearer to use the struct name in a non-generic context" would be helpful here.